pkg/apis/workflow/v1alpha1/workflow_types.go

```diff
@@ -1140,11 +1140,46 @@ type WorkflowStep struct {
 	// OnExit is a template reference which is invoked at the end of the
 	// template, irrespective of the success, failure, or error of the
 	// primary template.
+	// DEPRECATED: Use LifecycleHook.template instead.
 	OnExit string `json:"onExit,omitempty" protobuf:"bytes,11,opt,name=onExit"`
+
+	// Hooks holds exit hook which is invoked at the end of the
+	// template, irrespective of the success, failure, or error status of the primary template
+	Hooks *LifecycleHooks `json:"hooks,omitempty" protobuf:"bytes,12,opt,name=hooks"`
+}
+
+type LifecycleHooks struct {
```

Comment:
I know we're nearly there with this PR, but I will ask for a follow-up to changes this type to:

```
type LifecycleEvent string

const (
   LifecycleEvent = "Exit"
)

type LifecycleHooks map[Hook]LifecycleHook
```

That way we can generalise the code to support

```
LifecycleEvent = "Start"
```

I just want to give you a heads-up.

Reply:
done